### PR TITLE
feat/Cancel taskrun using entrypoint binary

### PR DESCRIPTION
**Pull request by chengjoey**
The cancellation of taskruns is now done through the entrypoint binary through a new flag called 'stop_on_cancel'. This removes the need for deleting the pods to cancel a taskrun, allowing examination of the logs on the pods from cancelled taskruns. Part of work on issue #3238 
there are some codes carry over from #4618 

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
if "enable-cancel-using-entrypoint" is set to
"true" in the `feature-flags`, Taskrun can now be cancelled using the entrypoint, which results of Pods from these TaskRun to be no longer deleted: they are stopped instead.
```

### CODE REVIEW

**Comment by chengjoey** [comment]
/hold
until add more tests

**Comment by chengjoey** [comment]
example taskrun.yaml:
```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: cancel
spec:
  taskSpec:
    steps:
    - name: step-1
      image: alpine
      script: echo "step-1 executing" && sleep 1200
```

after kubectl edit taskrun and add status `TaskRunCancelled`

check pod log:
image

describe pod:
image

**Comment by chengjoey** [comment]
https://github.com/tektoncd/pipeline/pull/4618#discussion_r862854824

I changed the logic of this part to `entrypointer` to judge whether the `ContextCanceld` error was triggered to determine whether the task was canceled

**Comment by chengjoey** [comment]
https://github.com/tektoncd/pipeline/pull/4618#discussion_r862859354

cancel or fail the taskrun has been merged into one function

**Comment by chengjoey** [comment]
https://github.com/tektoncd/pipeline/pull/4618#issuecomment-1115745946

For PipelineRun or TaskRun reconciler, the cancellation logic is just changed from delete pod to patch pod

**Review by vdemeester — APPROVED** [review]
cc @lbernick

**Comment by vdemeester** [comment]
@chengjoey this needs a rebase 😓 😅

**Comment by chengjoey** [comment]
> @chengjoey this needs a rebase 😓 😅

thanks @vdemeester , pr has been rebased

**Comment by chengjoey** [comment]
@jerop could you please review when you have time

**Comment by vdemeester** [comment]
@lbernick 
```
    waiter_test.go:307: expected Wait() to have detected a non-zero file size by now
```

File: cmd/entrypoint/main.go
Comment (lbernick) [inline cmd/entrypoint/main.go:189]: curious why SIGKILL is used here? could you add a comment about this?

File: cmd/entrypoint/runner_test.go
Comment (lbernick) [inline cmd/entrypoint/runner_test.go:214]: nit: would it make sense to run `rr.Run` in a goroutine instead and cancel the context? This would avoid the time delay before the test succeeds and may be less likely to be flaky.

File: pkg/entrypoint/entrypointer.go
Comment (lbernick) [inline pkg/entrypoint/entrypointer.go:117]: why not modify the existing Wait function to accept a context?

File: docs/taskruns.md
Comment (lbernick) [inline docs/taskruns.md:887]: Could we rename this feature flag in terms of the user-facing impact? e.g. "keep-pod-on-cancel" or something like that. Also, it's probably worth noting this uses the downward API, and might not be suitable in environments that don't allow it (https://github.com/tektoncd/pipeline/issues/4937). This is why we introduced the "await-sidecar-readiness" flag. We should also note the default value. lastly, it's probably worth describing this as an alpha feature, and adding it to our alpha features table. https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#alpha-features

File: pkg/apis/config/feature_flags_test.go
Comment (lbernick) [inline pkg/apis/config/feature_flags_test.go:73]: nit: can you remove this extra line please?

File: pkg/apis/config/feature_flags.go
Comment (lbernick) [inline pkg/apis/config/feature_flags.go:94]: Could you add this feature flag here please? https://github.com/tektoncd/pipeline/blob/main/config/config-feature-flags.yaml

File: pkg/entrypoint/entrypointer.go
Comment (lbernick) [inline pkg/entrypoint/entrypointer.go:60]: I'm curious if these error types are necessary given that they just wrap errors from the context package?

File: pkg/pod/entrypoint.go
Comment (lbernick) [inline pkg/pod/entrypoint.go:190]: Is it necessary to add this as an entrypoint arg, or can the entrypoint just listen for this file by default? The taskrun controller is already canceling in different ways based on the value of the feature flag.

File: pkg/pod/entrypoint.go
Comment (lbernick) [inline pkg/pod/entrypoint.go:256]: nit: I prefer not to include details of the calling function in a function docstring. also, it's still possible to inspect error types w/ wrapped errors using errors.is

File: pkg/pod/entrypoint.go
Comment (lbernick) [inline pkg/pod/entrypoint.go:267]: nit: since this is a one-line function call, is a separate function needed?

File: pkg/reconciler/taskrun/taskrun.go
Comment (lbernick) [inline pkg/reconciler/taskrun/taskrun.go:686]: nit: this may be simpler with if/else

File: cmd/entrypoint/main.go
Comment (chengjoey) [reply cmd/entrypoint/main.go:189]: use the SIGKILL signal to distinguish normal exit programs, just like kill -9 PID.

added comment

File: pkg/apis/config/feature_flags_test.go
Comment (chengjoey) [reply pkg/apis/config/feature_flags_test.go:73]: removed

File: pkg/apis/config/feature_flags.go
Comment (chengjoey) [reply pkg/apis/config/feature_flags.go:94]: added

File: cmd/entrypoint/runner_test.go
Comment (chengjoey) [reply cmd/entrypoint/runner_test.go]: removed this test

Comment (lbernick) [reply cmd/entrypoint/runner_test.go]: I think this is probably a useful test; I'd rather have it around without the suggestions I made than have it removed!

File: pkg/entrypoint/entrypointer.go
Comment (lbernick) [inline pkg/entrypoint/entrypointer.go:178]: Is it possible that this goroutine would leak if the taskrun is never canceled? i.e. the goroutine is waiting for a downwardapi file to exist; if it never does, I'm not sure whether the goroutine will terminate.

File: cmd/entrypoint/runner_test.go
Comment (chengjoey) [reply cmd/entrypoint/runner_test.go]: thanks @lbernick , I merged the two tests into `TestRealRunnerCancel`, and canceled at a random time, trying to test all the situations

File: docs/taskruns.md
Comment (chengjoey) [reply docs/taskruns.md]: 1. renamed to "keep-pod-on-cancel" and add this flag to alpha-features
2. at present, it is still necessary to use the downward API to pass the cancel signal. setting `keep-pod-on-cancel` to true will make the pod use the downward api despite using the "await-sidecar-readiness" flag

File: pkg/pod/entrypoint.go
Comment (chengjoey) [reply pkg/pod/entrypoint.go]: Good Idea! i removed the `stop_on_cancel`. feature flag and taskrun controller are enough to cancel tasks in different ways

File: pkg/pod/entrypoint.go
Comment (chengjoey) [reply pkg/pod/entrypoint.go]: adjusted

File: pkg/reconciler/taskrun/taskrun.go
Comment (chengjoey) [reply pkg/reconciler/taskrun/taskrun.go]: adjusted

File: pkg/entrypoint/entrypointer.go
Comment (chengjoey) [reply pkg/entrypoint/entrypointer.go:176]: Thank you for your reminder, I added defer cancel(), which will not affect the cancellation function, but also allow the goroutine to be released

**Review by jerop — COMMENTED** [review]
@chengjoey thank you for these contributions, and for the patience with the reviews

this pull request generally looks good to me, one ask is that you improve the code coverage to at least maintain the current coverage e.g. add a test case for setting the feature flag to something invalid

**Comment by chengjoey** [comment]
> @chengjoey thank you for these contributions, and for the patience with the reviews
> 
> this pull request generally looks good to me, one ask is that you improve the code coverage to at least maintain the current coverage e.g. add a test case for setting the feature flag to something invalid

Thanks~ i have added some test cases for `entrypoint` `feature-flags` `entrypointer`
PLAT @jerop @lbernick

**Comment by chitrangpatel** [comment]
@chengjoey Thanks for your work on this. The changes look good to me. Please rebase this PR. I'm happy to LGTM it after.

**Comment by chengjoey** [comment]
> @chengjoey Thanks for your work on this. The changes look good to me. Please rebase this PR. I'm happy to LGTM it after.

Thanks @chitrangpatel , pr has been rebased

**Comment by Yongxuanzhang** [comment]
I think this PR has introduced flaky unit tests:
https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/7167/pull-tekton-pipeline-unit-tests/1714307892095488000
https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/7204/pull-tekton-pipeline-unit-tests/1714305917400387584
https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/7167/pull-tekton-pipeline-unit-tests/1714304663337046016
https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/7193/pull-tekton-pipeline-unit-tests/1714303905107546112

**Pull request merged**
